python/ray/llm/_internal/batch/stages/vllm_engine_stage.py

```diff
@@ -0,0 +1,530 @@
+"""The stage that runs vLLM engine."""
+import asyncio
+import dataclasses
+import importlib
+import logging
+import math
+import os
+import time
+import uuid
+from dataclasses import dataclass
+from pydantic import root_validator
+from typing import TYPE_CHECKING, Any, Dict, AsyncIterator, Optional, List, Tuple
+
+import ray
+from ray.llm._internal.batch.stages.base import (
+    StatefulStage,
+    StatefulStageUDF,
+)
+from ray.util.scheduling_strategies import PlacementGroupSchedulingStrategy
+
+if TYPE_CHECKING:
+    from PIL import Image
+
+logger = logging.getLogger(__name__)
+
+
+class vLLMEngineWrapper:
+    """Wrapper around the vLLM engine to handle async requests.
+
+    Args:
+        *args: The positional arguments for the engine.
+        max_pending_requests: The maximum number of pending requests in the queue.
+        runtime_env: The runtime environment to use for the vLLM engine.
+        **kwargs: The keyword arguments for the engine.
+    """
+
+    @dataclass(frozen=True)
```

Comment:
why are you not defining this class outside of this class? I am allergic to in-class class or in-function function definitions.

Reply:
Because this is only used by this class and I'm not allergic to in-class class lol
No strong preference tho, I can move it out and make a more specific name like vLLMRequest.

Reply:
it impacts readability :D and opens door for overuse and weird bugs (using variables that are defined in scope and are overriding the parents scope, etc)